tests/kafkatest/tests/streams/streams_eos_upgrade_test.py

```diff
@@ -0,0 +1,255 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+from ducktape.mark import matrix
+from ducktape.mark.resource import cluster
+from kafkatest.tests.kafka_test import KafkaTest
+from kafkatest.services.streams import StreamsEosTestDriverService, StreamsEosTestJobRunnerService, \
+    StreamsComplexEosTestJobRunnerService, StreamsEosTestVerifyRunnerService, StreamsComplexEosTestVerifyRunnerService
+from kafkatest.version import LATEST_3_0, LATEST_3_1, LATEST_3_2, DEV_VERSION
+
+# TODO: Write upgrade noticex
+eos_v1_versions = [str(LATEST_3_0), str(LATEST_3_1), str(LATEST_3_2)]
+eos_v2_versions = [str(DEV_VERSION)]
+
+class StreamsEosUpgradeTest(KafkaTest):
+
+    processed_data_msg = "processed [0-9]* records from topic=data"
+    base_version_number = str(DEV_VERSION).split("-")[0]
+
+    """
+    Test of Kafka Streams exactly-once semantics
+    """
+
+    def __init__(self, test_context):
+        super(StreamsEosUpgradeTest, self).__init__(test_context, num_zk=1, num_brokers=3, topics={
+            'data': {'partitions': 5, 'replication-factor': 2},
+            'echo': {'partitions': 5, 'replication-factor': 2},
+            'min': {'partitions': 5, 'replication-factor': 2},
+            'sum': {'partitions': 5, 'replication-factor': 2},
+            'repartition': {'partitions': 5, 'replication-factor': 2},
+            'max': {'partitions': 5, 'replication-factor': 2},
+            'cnt': {'partitions': 5, 'replication-factor': 2}
+        })
+        self.driver = StreamsEosTestDriverService(test_context, self.kafka)
+        self.test_context = test_context
+
+    @cluster(num_nodes=9)
+    @matrix(from_version=eos_v1_versions,
+            to_version=eos_v2_versions)
+    def test_eos_upgrade_simple(self, from_version, to_version):
```

Comment:
To fill in the removed coverage from the deleted integration tests, could we also have a test for 3.3 (and potentially for 3.4 in the future) where we did not change the byte code version but just changing the eos config in a rolling bounce?

Reply:
Yeah I think you're right, never mind :)